Feature/aksyonov/3 game logic (#46)
333EYE666CORP999/MentalMath: Docs/2_Problem Generation.md

# Problem Generation

## Intro

This doc is focused on describing how problems are generated, brealing down the anatomy of generation process.

## Rationale

As discussed at the very beginning, all problem generation logic should be abstracted to separate namespace. This will allow for scalability so the mobile teams would only have to write frontend. Game aspects such as:

1. Problem generation
2. Problem solution check
3. Score tracking
4. Statistics
5. Analytics
6. Insights

will be stored in separate package written in [Rust](https://www.rust-lang.org). See technical consideration of selectig Rust as a primary language for game logic in [Core in Rust](./4_4_TDR%3A%20Core%20in%20Rust.md). Not only should the game logic be thoroughly encoded in a framework, it should have meaningful textual description to facilitate migration to Rust. Moreover, stated methodology mill allow for targetet unit-testing.

## Enforcing Methodology in Code

> `XCTestCase` is a good option to enforce methodology thtrough the system of test-calses. For one, randomness ratios, edge cases ans conforming to methodology principles can be testet thoughout unit-tests. Having these tests in place will alert a developer each time he has broken the crucial logic.

Author: ```suggestion
`XCTestCase` is a good option to enforce methodology through the system of test cases. For one, randomness ratios, edge cases and conforming to methodology principles can be tested via unit testing. Having these tests in place will alert a developer each time he has broken the crucial logic.
These tests should be structured according to the methodology described below. 
```suggestion

## Problem Generation Methodology

> In general, problem generation flow can be enumerated by the follwing steps:

Author: ```suggestion
Problem generation anatomy comprises of the following steps:
```

1. Determining the operation
2. Selecting digits capacity

> 3. Defining the series' upper bound for random number selection

Author: ```suggestion
3. Defining the upper and lower bounds for random number selection range
```

> 4. Pikcing random number from the series

Author: ```suggestion
4. Pikcing random number from the range
```
